pkg/apis/pipeline/v1alpha1/verificationpolicy_validation.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2022 The Tekton Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/tektoncd/pipeline/pkg/apis/validate"
+	"knative.dev/pkg/apis"
+)
+
+var _ apis.Validatable = (*VerificationPolicy)(nil)
+
+var (
+	// supportedSignatureAlgorithms sets a list of support signature algorithms that is similar to the list supported by cosign
+	supportedSignatureAlgorithms = []string{"sha224", "sha256", "sha384", "sha512"}
+)
+
+// Validate VerificationPolicy
+func (v *VerificationPolicy) Validate(ctx context.Context) (errs *apis.FieldError) {
+	errs = errs.Also(validate.ObjectMetadata(v.GetObjectMeta()).ViaField("metadata"))
+	errs = errs.Also(v.Spec.Validate(ctx))
+	return errs
+}
+
+// Validate VerificationPolicySpec, the validation requires Resources is not empty, for each
+// resource it must be able to be regex expression and can be compiled with no error. The Authorities
+// shouldn't be empty and each Authority should be valid.
+func (vs *VerificationPolicySpec) Validate(ctx context.Context) (errs *apis.FieldError) {
+	if len(vs.Resources) == 0 {
+		errs = errs.Also(apis.ErrMissingField("resources"))
+	}
+	for _, r := range vs.Resources {
+		errs = errs.Also(r.Validate(ctx))
+	}
+	if len(vs.Authorities) == 0 {
+		errs = errs.Also(apis.ErrMissingField("authorities"))
+	}
+	for i, a := range vs.Authorities {
+		if a.Key != nil {
+			errs = errs.Also(a.Key.Validate(ctx).ViaFieldIndex("key", i))
+		}
+	}
+	return errs
+}
+
+// Validate KeyRef will check if one of KeyRef's Data or SecretRef exists, and the
+// HashAlgorithm is one of "sha224", "sha256", "sha384", "sha512".
```

Comment:
Instead of listing the valid values in multiple places, just refer to the list of constants. (We're never going to keep the code in sync with the doc when more types are added!)